crates/matrix-sdk/src/client/mod.rs

```diff
@@ -677,20 +677,35 @@ impl Client {
         H: EventHandler<Ev, Ctx>,
         <H::Future as Future>::Output: EventHandlerResult,
     {
-        self.add_event_handler(handler).await;
+        self.add_event_handler(handler);
         self
     }
 
-    pub(crate) async fn event_handlers(&self) -> RwLockReadGuard<'_, EventHandlerMap> {
-        self.inner.event_handlers.read().await
+    pub(crate) fn event_handlers(&self) -> StdRwLockReadGuard<'_, EventHandlerMap> {
+        self.inner.event_handlers.read().unwrap()
     }
 
     /// Remove the event handler associated with the handle.
     ///
-    /// Note that handlers that remove themselves will still execute
-    /// with events received in the same sync cycle.
+    /// Note that you **must not** call `remove_event_handler` from the
+    /// non-async part of an event handler, that is:
+    ///
+    /// ```ignore
+    /// client.add_event_handler(|ev: SomeEvent, client: Client, handle: EventHandlerHandle| {
+    ///     // ⚠ this will cause a deadlock ⚠
```

Comment:
Will this cause a deadlock? We do the following
1. Lock the event handlers map to read the handlers
2. Construct the futures for the handler callbacks
3. Release the lock
4. Run the futures in order.

We will try to remove the handler at step 4 which shouldn't hold on to the lock anymore, no?

Reply:
No, this is step 2, below where there's another comment is step 4.

Reply:
Ah right, that's not part of the future, that's part of the method that constructs the future.